frontend/appflowy_flutter/lib/plugins/database_view/grid/application/row/row_detail_bloc.dart

```diff
@@ -71,36 +96,60 @@ class RowDetailBloc extends Bloc<RowDetailEvent, RowDetailState> {
 
   Future<void> _startListening() async {
     rowController.addListener(
-      onRowChanged: (cells, reason) {
-        if (!isClosed) {
-          add(RowDetailEvent.didReceiveCellDatas(cells.values.toList()));
+      onRowChanged: (cellMap, reason) {
+        if (isClosed) {
+          return;
         }
+        final allCells = cellMap.values.toList();
+        final visibleCells = allCells
+            .where(
+              (cellContext) =>
+                  cellContext.isVisible(
+                    showHiddenFields: state.showHiddenFields,
+                  ) &&
+                  !cellContext.fieldInfo.isPrimary,
+            )
+            .toList();
+        final numHiddenFields = allCells
+            .where(
+              (cellContext) =>
+                  !cellContext.isVisible() && !cellContext.fieldInfo.isPrimary,
+            )
+            .length;
```

Comment:
Can we do this in one loop instead of two?
```suggestion
        int numHiddenFields = 0;
        List<DatabaseCellContext> visibleCells;
        for (final cell in allCells) {
          final isVisible = cell.isVisible();
          final isPrimary = cell.fieldInfo.isPrimary;
          
          if (state.showHiddenFields || (isVisible && !isPrimary)) {
            visibleCells.add(cell);
          }
          
          if (!isVisible && !isPrimary) {
            numHiddenFields++;
          }
        }
```